src/execution_algorithms/ltfb/regularized_evolution.cpp

```diff
@@ -0,0 +1,282 @@
+#include <vector>
+
+namespace lbann {
+namespace ltfb {
+
+// RegularizedEvolution Implementation
+
+RegularizedEvolution::RegularizedEvolution(
+  std::string metric_name,
+  metric_strategy winner_strategy,
+  std::unique_ptr<MutationStrategy> mutate_algo,
+  int sample_size)
+  : m_metric_name{std::move(metric_name)},
+    m_metric_strategy{std::move(winner_strategy)},
+    m_mutate_algo{std::move(mutate_algo)},
+    m_sample_size{std::move(sample_size)}
+{}
+
+RegularizedEvolution::RegularizedEvolution(RegularizedEvolution const& other)
+  : m_metric_name{other.m_metric_name},
+    m_metric_strategy{other.m_metric_strategy},
+    m_mutate_algo{other.m_mutate_algo->clone()},
+    m_sample_size{other.m_sample_size}
+{}
+
+EvalType RegularizedEvolution::evaluate_model(model& m,
+                                              ExecutionContext& ctxt,
+                                              data_coordinator& dc) const
+
+{
+  // Make sure data readers finish asynchronous work
+  const auto original_mode = ctxt.get_execution_mode();
+  dc.collect_background_data_fetch(original_mode);
+
+  // Can use validation if it is global
+  if (!dc.is_execution_mode_valid(execution_mode::tournament)) {
+    LBANN_ERROR("Regularized Evolution requires ",
+                to_string(execution_mode::tournament),
+                " execution mode");
+  }
+
+  // Mark the data store as loading - Note that this is a temporary fix
+  // for the current use of the tournament
+  m.mark_data_store_explicitly_loading(execution_mode::tournament);
+
+  // Evaluate model on test (or validation?) set
+  get_trainer().evaluate(&m, execution_mode::tournament);
+
+  // Get metric values
+  bool found_metric = false;
+  EvalType score = 0.f;
+  std::string metric_name;
+  for (const auto& met : m.get_metrics()) {
+    metric_name = met->name();
+    if (metric_name == m_metric_name) {
+      found_metric = true;
+      score += met->get_mean_value(execution_mode::tournament);
+      break;
+    }
+  }
+
+  // sanity check
+  if (!found_metric) {
+    LBANN_ERROR("could not find metric \"",
+                metric_name,
+                "\" "
+                "in model \"",
+                m.get_name(),
+                "\"");
+  }
+
+  m.make_data_store_preloaded(execution_mode::tournament);
+
+  // Clean up and return metric score
+  m.reset_mode(ctxt, original_mode);
+  dc.reset_mode(ctxt);
+  return score;
+}
+
+void RegularizedEvolution::select_next(model& m,
+                                       ltfb::ExecutionContext& ctxt,
+                                       data_coordinator& dc) const
+{
+  auto const& comm = *(m.get_comm());
+  const unsigned int num_trainers = comm.get_num_trainers();
+  const int trainer_id = comm.get_trainer_rank();
+  auto const step = ctxt.get_step();
+
+  std::vector<EvalType> sample_trainers(num_trainers);
+  if (comm.am_world_master()) {
+    for (unsigned int i = 0; i < num_trainers; i++)
+      sample_trainers[i] = i;
```

Comment:
Why is an index vector using `EvalType` (i.e., `double`)?
```suggestion
  std::vector<unsigned> sample_trainers(num_trainers);
  if (comm.am_world_master()) {
    std::iota(begin(sample_trainers), end(sample_trainers), 0U);
```